Gip 0080 (#48)
graphprotocol/graph-improvement-proposals: gips/0080-gip-direct-indexing-payments.md

---
GIP: "0080"

> Title: <The title of this GIP>

Reviewer: ```suggestion
Title: A New Proposal for Indexing Fees
```

Authors: Tim Fong <tim@edgeandnode.com>, Pablo <pablo@edgeandnode.com>

> Created: <The date the GIP was created.>

Reviewer 2: ```suggestion
Created: <2024-08-22>
```

> Updated: <The date the GIP was last updated.>

Reviewer 2: ```suggestion
Updated: <2024-08-22>
```

Stage: <The current stage of the proposal. Specified by the author and confirmed by editors by virtue of a GIP being accepted into an editor's view of the repo.>
Discussions-To: <The forum post where this proposal is discussed.>
Category: <(Optional) The type of GIP or GRP. This field should be left blank for GRCs. Valid types are "Protocol Logic", "Protocol Interfaces", "Subgraph API", "Process", "Economic Parameters", and "Protocol Charters".>
Depends-On: <(Optional) A list of GIPs that this GIP depends on. If some other type of dependency exists, include a reference link here and an explanation in the body of the GIP.>
Superseded-By: <(Optional) A GIP that supersedes the current proposal. If this field is specified, the stage of the GIP should be "Withdrawn".>

> Replaces: <(Optional) A GIP that this proposal is intended to supersede.>

Reviewer: ```suggestion
Replaces: GIP-0058
```

Resolves: <(Optional) If this GIP was written in response to an RFP, include it here.>
Community-Polls: <(Optional) A list of URLs to community polls relating to this GIP.>
Governance Proposals: <(Optional) A list of URLs to governance proposals related to this GIP.>
Implementations: <(Optional) A list of URLs to reference implementations for this proposal.>
Audits: <(Optional) A list of URLs to audits related to this proposal.>
---

## Abstract

> This supercedes GIP-0058, which introduced a protocol mechanism for Gateways to pay Indexers directly for indexing subgraphs.  This updated variation doesn’t introduce new major concepts, but does the following: describes the payment mechanism with more detail, provides a more detailed taxonomy of the participants, and provides a comparison of the Protocol versus the Product.  Enabling Indexing Fees provides an alternative alongside the current Curation model.

Reviewer: nit: supercedes -> supersedes

> This supercedes GIP-0058, which introduced a protocol mechanism for Gateways to pay Indexers directly for indexing subgraphs.  This updated variation doesn’t introduce new major concepts, but does the following: describes the payment mechanism with more detail, provides a more detailed taxonomy of the participants, and provides a comparison of the Protocol versus the Product.  Enabling Indexing Fees provides an alternative alongside the current Curation model.

Reviewer: ```suggestion
This supersedes GIP-0058, which introduced a protocol mechanism for Gateways to pay Indexers directly for indexing subgraphs. This proposal is a simplified version of the same ideas, presenting it as an alternative alongside the current Curation and Rewards model. It describes the payment mechanism with more detail, provides a more detailed taxonomy of the participants, and provides a distinction between the Protocol and its related Products.
```

This GIP isn’t specifying execution timing, but introduces two high-level phases: a) an off-chain MVP; b) a post-MVP contracts-based implementation. 

## Motivation

> Currently, subgraph developers use the Curation mechanism to attract Indexers to their subgraph. Curation adds friction by locking a potentially high amount of GRT, and delivers a low predictability for the required amount of GRT needed to achieve a certain Quality of Service (QoS). The friction from Curation may make it harder for developers to select The Graph over alternatives.

Reviewer 2: ```suggestion
Currently, subgraph developers use the curation mechanism to attract Indexers to their subgraph. curation adds friction by locking a potentially high amount of GRT, and delivers a low predictability for the required amount of GRT needed to achieve a certain Quality of Service (QoS). The friction from curation may make it harder for developers to select The Graph over alternatives.
```

> Moreover, the current Indexing Rewards mechanism does not reward Indexers  for serving queries with a high QoS, and does not provide Indexers with predictability on the revenue they can make for indexing and serving subgraphs.

Reviewer 2: ```suggestion
Moreover, the current indexing rewards mechanism does not reward Indexers for serving queries with a high QoS, and does not provide Indexers with predictability on the revenue they can make for indexing and serving subgraphs.
```

## High-Level Description
There are two primary phases: off-chain MVP and Post-MVP contracts-based implementation.  Here are the high-level components of each phase.

### Off-Chain MVP
The MVP implementation can be done without any changes to the protocol smart contracts. Interactions between Gateways and Indexers happen through HTTP APIs, and payments are sent as a lump sum as if it were a single query fee payment, to be collected using rebates together with the query fees for the allocation.

> This allows the development of the simplest possible implementation with the existing stack, though at this stage Indexers need to trust that Gateways will pay after the POI is posted. For simplicity, the MVP will use an indexing price as a function of entities and indexed blocks.

Reviewer: ```suggestion
This allows the development of the simplest possible implementation with the existing stack, though at this stage Indexers need to trust that Gateways will pay after the POI is posted. For simplicity, the MVP will use an indexing price as a function of entities and indexed blocks, which could be different for each chain.
```

The following is a sequence diagram

![image](https://github.com/user-attachments/assets/1182cc24-9adf-4aea-969a-954b9364389f)

### Post-MVP Contracts-Based Implementation

> The Post-MVP phase depends on the Subgraph Service implementation from Graph Horizon, and uses a trust-minimized payment mechanism for Indexers to collect payment from an Escrow (that is the same Escrow used by TAP for query fee payments). 

Reviewer: ```suggestion
The Post-MVP phase depends on the Subgraph Service implementation from Graph Horizon (GIP-0066 and GIP-0068), and uses a trust-minimized payment mechanism for Indexers to collect payment from an Escrow (that is the same Escrow used by TAP for query fee payments). 
```

The Gateway commits to paying the specified price with a voucher, that is posted on-chain by the Indexer, and that the contracts will use as a pre-authorization. 

This allows the Indexer to collect payment directly when posting the POI, and having a guarantee that the payment will be available when the job is done (as long as the Escrow is sufficiently funded). 

At this phase, we will likely also iterate on the pricing approach and introduce a better "indexing gas" metric as an evolution to the MVP entities-and-blocks-based approach. 

### Product vs Protocol

> The GIP is primarily enabling Consumers, who most likely will be Gateways, to interact via the protocol with the Indexer Network.

Reviewer: ```suggestion
The GIP is primarily enabling consumers at the protocol level, who most likely will be Gateways, to interact via the protocol with the Indexer Network.
```

> However, the actual Products which are the interface for Developers and End Customers, need a way to successfully build their business ontop of the Protocol.

Reviewer 2: ```suggestion
However, the actual products which are the interface for developers and end customers, need a way to successfully build their business ontop of the protocol.
```

This illustration shows the relationship between the two:

![image](https://github.com/user-attachments/assets/0415f01d-4b8f-4bf0-9579-72acbe037a51)

### What is a Product?

> The Product Experience **defines the UX for the End User**, in most cases, a dapp developer who wants data that has been indexed.

Reviewer 2: ```suggestion
The product experience **defines the UX for the End User**, in most cases, a dapp developer who wants data that has been indexed.
```

> A good Product experience includes as much **predictability and transparency upfront for potential costs to the paying Developer.**

Reviewer 2: ```suggestion
A good product experience includes as much **predictability and transparency upfront for potential costs to the paying developer.**
```

However, there could be instances the Gateway has to take on risks for the sake of simplicity.

**This decision is a Product decision** made by the Studio operator.

### What is the role of the Gateway in the Product experience?

The Studio operator will often be running the Gateway, as well (although if a trusted experience is established with a third-party Studio, this could also be possible).

> The Gateway, then, is like the payment gateway **and** the Indexing Indexer Selection Algorithm (IISA) for the Studio operator.  As a payment gateway, the Gateway operates more like Stripe: a utility facilitating the payment flow via escrow.  However, the IISA **could** be considered part of the product experience, a means for the Studio to potentially differentiate in the marketplace.

Reviewer: The comparison with Stripe here seems a bit odd, I'd suggest removing it

Could another Studio use the same Gateway?

> If the Studio used an existing Studio via an API, such as a Studio-as-a-Service”, then another Studio would inherit the trust assumptions of the Gateway operator (since the Studio and Gateway are the same operator). 

Reviewer: This sentence looks a bit confusing - the studio using an existing studio? imo it's not necessary to go into so much detail about things on the Product side, in the end GIPs are more about changes to the Protocol, and I think we only need to mention that there will be UX improvements that are possible on the Product side but won't directly affect the protocol design

Could a Studio choose to work with multiple Gateways?

Yes, there would need to be a trusted relationship. But a Studio which wanted to offer multiple data services and not operate the corresponding Gateway for each data service, they could. 

In this case, the Studio pays the Gateway (or the Gateway bill the Studio), such that a flow of funds goes via the Escrow to the selected Indexing Indexers to fulfill the Agreement obligations for Indexing Network Payment.

## Detailed Specification

> ### The payments mechanism

Reviewer: ```suggestion
### Rationale for the payments mechanism
```

**This voucher release mechanism is generic and can be used for other data services:** it validates that the caller is the approved data service contract, that the token amount is within the maximum value specified by the voucher per period, and that the right amount of time has passed. The data service specific aspects e.g. POI validity, gas/pricing, etc. are defined at the data service contract level. Moreover, this contract can be immutable, so payers are protected even if the data service contract is upgraded.

It is important to note that we cannot expect Indexers to take on risk, e.g. by agreeing to a fixed price for a specific subgraph (instead of per unit of work/gas). The cost of indexing each subgraph is unknown until the subgraph has been indexed, so if the price was fixed a rational Indexer would simply stop indexing when the subgraph becomes unprofitable, forcing the payer to start over. Tying payments as much as possible to the verifiable cost of performing the work is the best way to prevent Denial of Service scenarios, just like with gas in Ethereum.

### Terms and Definitions

> **Curation**: Curation requires Consumers to stake a certain amount of GRT per subgraph to signal Indexers to provide indexing services.  However, because the strength of the signal depended upon the strength of other signals using GRT, which also had variability, this became a hard mechanism for developers with many subgraphs.

Reviewer: ```suggestion
**Curation**: Curation requires Consumers to "signal" a certain amount of GRT (by depositing it to a smart contract) per subgraph to signal Indexers to provide indexing services, as curation signal affects the amount of Indexing Rewards.  However, because the strength of the signal depends upon the amount of curation on other subgraphs, which has variability, this can be a hard mechanism for developers with many subgraphs.
```

**Indexers:** Indexers are node operators in The Graph Network that stake Graph Tokens (GRT) in order to provide indexing and query processing services. Indexers earn query fees and indexing rewards for their services.

**Indexing Network Payments:**  this represents the payments to Indexers upon verifying successful indexing.  Typically will occur between a Gateway and Indexers selected by their IISA.

**End Consumers:** The core users of data services via Products built onto the Protocol are referred to as Consumers. Specifically, the consumer is the one that receives a benefit (and thus is willing to pay) for data services. In some cases, this may be the developer. In other cases, a consumer may want to leverage the subgraph of another developer. In general, the consumer is the party that interacts with Gateways/Indexers for the indexing services.

> **Consumer**: This is referring to one who interacts directly with the Protocol for indexing services.  Often a **Gateway** (see below).

Reviewer: I think this can cause a bit of confusion with the "End Consumer" above. Maybe we should distinguish "End Consumers" or "Consumers at the Product level" from "Consumers at the Protocol level, usually Gateways"?

**Gateway:** a permissionless entity which can select an Indexer and pay for Indexing; often associated with a **Studio**

**Studio:** the primary "Product" that End Customers or developers interact with to pay for Indexing and to query Indexed Subgraphs

**Indexing Gas or Unit of Work:** a concept still being defined that will be used to measure the work required to deliver the indexing.  This is likely a combination of compute, storage, and network related “work” needed to be performed by Indexers providing the indexing service.  This could potentially vary by chain.

> **Indexing Gas:** this will be the posted “price” by an Indexer in order to perform indexing services based on the unit of work..

Reviewer: This is repeated from above, and the Indexing Gas is not the posted price, but the unit that is offered for a price

> **Horizon Escrow:** to trust-minimize the system, as part of the Horizon roll-out, the Gateway will place payments from the Consumer into Horizon Escrow.  The Indexer will be paid from Horizon.  However, there will be a “trusted” version released without Horizon Escrow.

Reviewer: In the current Horizon code we're calling this "GraphEscrow"

Reviewer: also, this is not accurate: "Gateway will place payments from the Consumer into Horizon Escrow" - the Gateway will place its own funds on the Escrow - this may or may not be directly related to payments from end consumers

> **Agreement**: An agreement is used to identify the terms of service between an Indexer and the Gateway.  The negotiations typically take place off-chain, and formalized on-chain.  This Agreement can occur between a Gateway and multiple Indexers, and has some concept of cost.  Typically encoded as a **Voucher**.

Reviewer: ```suggestion
**Agreement**: An agreement is used to identify the terms of service between an Indexer and the Gateway.  The negotiations typically take place offchain (in an automated manner), and formalized onchain.  A Gateway can create Agreements for the same subgraph with several Indexers. The terms of service are encoded in a **Voucher** signed by the Gateway, and the Indexer accepts the Agreement by posting the Voucher onchain..
```
